templates/support/OAuthAccessTokenManager.swift.hbs

```diff
@@ -0,0 +1,126 @@
+//  
+//  {{>frag/generatedBy}}
+//
+
+/** A token manager that will group refresh attempts to prevent concurrent token requests */
+import Foundation
+
+actor OAuthAccessTokenManager {
+    
+    var accessToken: OAuthToken? {
+        didSet {
+            restartRefreshTimer()
+        }
+    }
+    
+    private let refreshTokenURL: URL?
+    private let automaticallyRefreshTokensThisManySecondsEarly: TimeInterval?
+    private var refreshTask: Task<(), Error>?
+    private var refreshTimer: Timer?
+    private let prepareRequest: ((_ request: URLRequest) -> URLRequest)?
+    
+    init(refreshTokenURL: URL?, automaticallyRefreshTokensThisManySecondsEarly: TimeInterval? = nil, prepareRequest: ((_: URLRequest) -> URLRequest)? = nil) {
+        self.automaticallyRefreshTokensThisManySecondsEarly = automaticallyRefreshTokensThisManySecondsEarly
+        self.prepareRequest = prepareRequest
+        self.refreshTokenURL = refreshTokenURL
+    }
+        
+    private func restartRefreshTimer() {
+        refreshTimer?.invalidate()
+        refreshTimer = nil
+        
+        guard 
+            let autoTime = automaticallyRefreshTokensThisManySecondsEarly,
+            let token = accessToken, 
+            let expiresIn = token.expiresIn,
+            expiresIn - autoTime > 0
+        else {
+            return
+        }
+        
+        let timeout = Date(timeIntervalSinceNow: expiresIn - autoTime)
+        refreshTimer = Timer.scheduledTimer(withTimeInterval: expiresIn - autoTime, repeats: false, block: { t in
+            Task {
+                do {
+                    try await self.refreshToken()
+                } catch {
+                    print("An error ocurred in the automatic refresh: \(error)")
+                }
+            }
+        })
+    }
+    
+    func setAccessToken(_ token: OAuthToken) async throws {
+        accessToken = token
+    }
+    
+    func setRefreshToken(_ token: String) async throws {
+        accessToken = OAuthToken(refreshToken: token)
+    }
+    
+    func refreshToken() async throws {
+        guard let refreshTokenURL = refreshTokenURL else {
+            throw APIError.authenticationPrevented
```

Comment:
It's almost like here we're simply saying "you're no longer authenticated" ? Maybe we need to look at each place where this func is called and see what a failure here means to that calling method.

Reply:
I'm not sure the distinction is necessary? if we don't have a url then we can't refresh, that doesn't mean we are no longer authenticated unless you also have an expired token (which obviously would be the common path) but we can refresh pre-emptively (or try to) and we'd get this error.